File: summary/go/3_quick_sort.go
package main

import (
	"fmt"
	"math/rand"
)

func quickSort(arr []int) []int {
	if len(arr) <= 1 {
		return arr
	}

	pivot := arr[rand.Intn(len(arr))]
	var less, eq, greater []int
	for _, v := range arr {
		switch {
		case v < pivot:
			less = append(less, v)
		case v == pivot:
			eq = append(eq, v)
		case v > pivot:
			greater = append(greater, v)
		}
	}

	less = quickSort(less)
	greater = quickSort(greater)

	return append(append(less, eq...), greater...)
}

func main() {
	var n int
	fmt.Scan(&n)
	arr := make([]int, n)
	for i := 0; i < n; i++ {
		fmt.Scan(&arr[i])
	}

	sortedArr := quickSort(arr)
	fmt.Println(sortedArr)
}

File: summary/elixir/3_quick_sort.exs
defmodule QuickSort do
  def sort([]), do: []
  def sort([pivot | rest]) do
    {less, greater} = Enum.split_with(rest, &(&1 <= pivot))
    sort(less) ++ [pivot] ++ sort(greater)
  end
end

# 例としてリストを用意してソートを試してみます
unsorted_list = [3, 1, 4, 1, 5, 9, 2, 6, 5, 3, 5]
sorted_list = QuickSort.sort(unsorted_list)
IO.inspect(sorted_list)
